DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/svd/src/svd_testcase.hpp

```diff
@@ -0,0 +1,212 @@
+        cur_diff = abs(mat_i[i]);
+      }
+
+      if (cur_diff > max_diff) max_diff = cur_diff;
+    }
+    return max_diff;
+  }
+
+  template <unsigned k_fixed_iteration = FIXED_ITERATIONS,
+            unsigned k_raw_latency = 110, int k_zero_threshold_1e = -8>
+  T run_test(sycl::queue q, int benchmark_rep = 1, bool print_matrices = false) {
+    std::vector<T> flat_A = col_major_A();
+    std::vector<T> flat_U(rows_A * rows_A);
+    std::vector<T> flat_S(rows_A * cols_A);
+    std::vector<T> flat_V(cols_A * cols_A);
+    std::vector<ac_int<1, false>> rank_deficient(1);
+
+    std::cout << "Running SVD test with input size " << rows_A << " x "
+              << cols_A << ", repeating " << benchmark_rep << " time(s)"
+              << std::endl;
+
+    delta_time =
+        singularValueDecomposition<rows_A, cols_A, k_fixed_iteration,
+                                   k_raw_latency, k_zero_threshold_1e, T>(
+            flat_A, flat_U, flat_S, flat_V, rank_deficient, q, benchmark_rep);
+
+    throughput = (benchmark_rep / delta_time);
+
+    compare_S(extract_singular_value(flat_S));
+
+    A_error = check_USV(flat_A, flat_U, flat_S, flat_V);
+    U_orthogonal_error = check_orthogonal(flat_U, rows_A, rows_A);
+    V_orthogonal_error = check_orthogonal(flat_V, cols_A, cols_A);
+
+    if (print_matrices) {
+      std::cout << "S:\n";
+      // print_matrix(extract_singular_value(flat_S), 1, rows_A);
+      svd_testbench_tool::print_matrix<T>(flat_S, rows_A, cols_A, true);
+      std::cout << "V:\n";
+      svd_testbench_tool::print_matrix<T>(flat_V, cols_A, cols_A, true);
+      std::cout << "U:\n";
+      svd_testbench_tool::print_matrix<T>(flat_U, rows_A, rows_A, true);
+      std::cout << "Rank deficient input: " 
+        << (rank_deficient[0] ? "True" : "False") << std::endl;
+    }
+    return std::max({S_error, A_error, U_orthogonal_error, V_orthogonal_error});
+  }
+
+  void print_result() {
+    std::cout << "Singular value error: " << S_error << "(" << S_error_r * 100
+              << "%)" << std::endl;
+    std::cout << "Decomposition error (A = USVt): " << A_error << "("
+              << A_error_r * 100 << "%)" << std::endl;
+    std::cout << "U orthogonal error: " << U_orthogonal_error << std::endl;
+    std::cout << "V orthogonal error: " << V_orthogonal_error << std::endl;
+    std::cout << "Total duration: " << delta_time << "s" << std::endl;
+    std::cout << "Throughput: " << throughput * 1e-3 << "k matrices/s"
+              << std::endl;
+  }
+};
+
+SVDTestcase<float, 4, 4> small_4x4(
+    std::vector<std::vector<float>>{
+        {0.47084338, 0.99594452, 0.47982739, 0.69202168},
+        {0.45148837, 0.72836647, 0.64691844, 0.62442883},
+        {0.80974833, 0.82555856, 0.30709051, 0.58230306},
+        {0.97898197, 0.98520343, 0.40133633, 0.85319924}},
+    std::vector<float>({2.79495619, 0.44521050, 0.19458290, 0.07948970}));
+
+SVDTestcase<float, 16, 8> testcase_16x8(
+    std::vector<std::vector<float>>{
```

Comment:
You can disable clang-format for these matrices so they can be formatted nicely:

https://clang.llvm.org/docs/ClangFormatStyleOptions.html#disabling-formatting-on-a-piece-of-code